### DISCUSSION TITLE
edge_bfs function iterates over all edge problems

### DISCUSSION THREAD

**Discussion by q1165600785** [networkx/networkx#7646] [Q&A] [answered]
![tulunpicture](https://github.com/user-attachments/assets/dd522b4b-95a4-4fa4-9f85-a9146aba7df6)
The traversal result I want is A-B is 1, B-C is 3, B-E is 2, and C-D is 4. However, the result given by edge_bfs is A-B is 1, B-C is 2, B-E is 3, and C-D is 4. edge_bfs will give priority to traversing the edges with more connections. I don't know if I can explain my problem clearly. Can netwokx achieve the traversal I want? I also uploaded a picture. I don't know if you can see it.

**Comment by dschult** [comment, marked as answer]
Actually `edge_bfs` just takes the next edge based on the order the edges are stored in the graph (which in turn is based on the order that nodes and edges are added to the graph -- edges are reported based on adjacency dicts keyed by node, so they don't always get reported in the order they are added, but nodes are reported in the order they are added and edges are reported based on that ordering of the nodes).

You might take a look at [generic_bfs_edges](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.traversal.breadth_first_search.generic_bfs_edges.html) which has an input parameter `neighbors` which takes an optional function with node as input and returns an iterator of the neighboring nodes in whatever order you want.  For example, `neighbors=sorted` will sort the neighbors so that your BFS explored neighbors in a sorted order.

Similarly `bfs_edges` has a similar argument for sorting the neighbors.

The `bfs_edges` functions only explore edges that build a tree (to find a spanning tree). `edges_bfs` reports all edges by including edges that make a cycle. But that function does not have a "sort_neighbors" style of argument. I don't think it would be hard to implement that feature using the examples from the `breadth_first_search.py` suite of functions. We'd be interested in such a feature.

**Reply by q1165600785** [reply]
Thank you for your answer, I understand what you mean. My graph is a directed graph. When traversing all edges in bfs_edges, I want to sort the edges of the same node. First, I want to get the edges with less connectivity (shallow edges), and then get the edges with more connectivity (deepest edges). I don't consider the order of these edges. How can I achieve this sorting? I need the help of a big shot

**Comment by q1165600785** [comment]
I thought about my question again today. I am sure that what I want to express is that when traversing all the edges in breadth first, I want to sort the depth of all the edges of a certain node. It seems that I need to do a depth first traversal again, first get the edge with the smallest depth, and finally get the edge with the largest depth. I don't know if my expression is clear. Can you please ask if this algorithm can be implemented?

**Reply by dschult** [reply]
It sounds like you will need to define what you mean by "depth" more precisely.
BTW the code for BFS and DFS is pretty short and easy to copy/edit/try.
That might help you figure out how to define the sort pattern you want for the edges incident to a node.
Take a look at the 20 lines of code in [generic_bfs_edges](https://github.com/networkx/networkx/blob/0aaa392fc0a9b2c672193850b6b518bb7b6de0c2/networkx/algorithms/traversal/breadth_first_search.py#L85).
It sounds like you need to focus on what you mean by "depth of the edges" and "less connectivity"

**Reply by q1165600785** [reply]
Thank you again for your help. Today, I will go back to some basic knowledge of graph theory. I think the depth I want to express is actually Graph.degree. When traversing the edges of a node, the node with the highest degree at the end of the edge has the deepest depth, and the node with the lowest degree has the shallowest depth. I don't know if I have expressed it clearly

**Comment by dschult** [comment]
To sort the neighbors by degree with `bfs_edges`, use the argument 
`sort_neighbors=lambda node : sorted(G[node], key=G.degree[nbr])`
If that is slow, define the function in a function with `degrees=dict(G.degree)` precomputed.
Something like:
```python
    degrees = dict(G.degree)
    def degree_sorted_nbrs(node):
        return sorted(G[node], key=degree[nbr])
    # code that calls bfs_edges
```
Note that this will not work for `edges_bfs` because it doesn't have this sorting functionality.
